### ISSUE TITLE
jax.optix vs jax.experimental.optimizers

### ISSUE THREAD

**Issue by KristianHolsheimer** [google/jax#2420]
Hi there,

I'm still relatively new to JAX and I was wondering what the difference is between the `jax.optix` and `jax.experimental.optimizers` modules?

Is there some added benefit in using one vs the other or is this just a redundancy in the API?

(if it's an API redundancy, perhaps it might be good to nip this in the bud so as to avoid some of the tf pitfalls)

**Comment by mtthss** [comment]
I am the author of `optix`. The library was designed to support composable gradient trasformations that allow to easily build custom optimisers. Among the key objectives were:

A) avoid monolytic optimisers as in TF and `experimental.optimizers` (for instance, avoiding arbitrary constraints such as being able to clip gradients and then apply adam, but not the opposite). This is achieved by exposing a collection of common gradient transformations, that can be combined (and combined with any other custom ones) using `optix.chain`.

B) support combining updates from separate optimisers operating on shared weights (this is common in multi-task settings where you often want to hold per task optimiser statistics). This is achieved by the API separating gradient transformations (that can be arbitrarily composed) from parameter updates (which is done at the end with `optix.apply_updates`).

Additionally, compared to `jax.experimental.optimizers` optix also avoids hiding `params` in the optimiser, this is useful both for A) and B). 

I agree having two optimisers APIs in core JAX is not the best. There is a decision here to be made by the JAX team, but personally looking forward I think we should have the following steps: 

1) extend the `optix` API so that the `update_fns` have the signature `update_fn(updates, state, params=None) -> new_updates` rather than the current `update_fn(updates, state) -> new_updates`. This is not needed in the most popular deep learning optimisers, but it is used in some more recent ones (e.g. Fromage). 

2) Once this is done I think we should just deprecate `jax.experimental.optimizers`.

Of course I realise that I am somewhat opinionated in this regard, but I do believe that A) and B) are fundamental problems in optimisers in TF and `jax.experimental.optimizers`, and we should not have them being inherited in an optimiser library that lives in the core JAX library. 

We could also move `both` optimisers out of core JAX, as separate libraries, and let people pick what they find most suited to their use case.

**Comment by KristianHolsheimer** [comment]
Thanks @mtthss that does clarify things a lot.

My 2 cents is that it's good to have optix as part of the core library. Perhaps there is room for high-level optimizers (those that conceal there `params`), but yeah, I agree that it might be better to move them to a separate library. If only to avoid confusion.

Also, great job on optix. It's great to have this composability in the optimizers, really cool.

**Comment by mtthss** [comment]
FYI @mattjj
